rust-runtime/aws-smithy-http-server/src/request.rs

```diff
@@ -0,0 +1,204 @@
+    pub fn uri(&self) -> &Uri {
+        &self.uri
+    }
+
+    /// Gets a mutable reference to the request URI.
+    pub fn uri_mut(&mut self) -> &mut Uri {
+        &mut self.uri
+    }
+
+    /// Get the request HTTP version.
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    /// Gets a mutable reference to the request HTTP version.
+    pub fn version_mut(&mut self) -> &mut Version {
+        &mut self.version
+    }
+
+    /// Gets a reference to the request headers.
+    ///
+    /// Returns `None` if the headers has been taken by another extractor.
+    pub fn headers(&self) -> Option<&HeaderMap> {
+        self.headers.as_ref()
+    }
+
+    /// Gets a mutable reference to the request headers.
+    ///
+    /// Returns `None` if the headers has been taken by another extractor.
+    pub fn headers_mut(&mut self) -> Option<&mut HeaderMap> {
+        self.headers.as_mut()
+    }
+
+    /// Takes the headers out of the request, leaving a `None` in its place.
+    pub fn take_headers(&mut self) -> Option<HeaderMap> {
+        self.headers.take()
+    }
+
+    /// Gets a reference to the request extensions.
+    ///
+    /// Returns `None` if the extensions has been taken by another extractor.
+    pub fn extensions(&self) -> Option<&Extensions> {
+        self.extensions.as_ref()
+    }
+
+    /// Gets a mutable reference to the request extensions.
+    ///
+    /// Returns `None` if the extensions has been taken by another extractor.
+    pub fn extensions_mut(&mut self) -> Option<&mut Extensions> {
+        self.extensions.as_mut()
+    }
+
+    /// Takes the extensions out of the request, leaving a `None` in its place.
+    pub fn take_extensions(&mut self) -> Option<Extensions> {
+        self.extensions.take()
+    }
+
+    /// Gets a reference to the request body.
+    ///
+    /// Returns `None` if the body has been taken by another extractor.
+    pub fn body(&self) -> Option<&B> {
+        self.body.as_ref()
+    }
+
+    /// Gets a mutable reference to the request body.
+    ///
+    /// Returns `None` if the body has been taken by another extractor.
+    pub fn body_mut(&mut self) -> Option<&mut B> {
+        self.body.as_mut()
+    }
+
+    /// Takes the body out of the request, leaving a `None` in its place.
+    pub fn take_body(&mut self) -> Option<B> {
+        self.body.take()
+    }
+}
+
+#[async_trait]
+impl<T, B> FromRequest<B> for Option<T>
+where
+    T: FromRequest<B>,
+    B: Send,
+{
+    type Rejection = Infallible;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Option<T>, Self::Rejection> {
+        Ok(T::from_request(req).await.ok())
+    }
+}
+
+#[async_trait]
+impl<T, B> FromRequest<B> for Result<T, T::Rejection>
+where
+    T: FromRequest<B>,
+    B: Send,
+{
+    type Rejection = Infallible;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        Ok(T::from_request(req).await)
+    }
+}
```

Comment:
Why do we need these? Shouldn't we only need operation inputs to implement `FromRequest`?